uboot build fail upto date Arch

The build of uboot is failing for me, any hints would be appreciated,

cp -u arm-trusted-firmware/build/rk3328/release/bl31/bl31.elf u-boot/bl31.elf
source u-boot/../u-boot-venv/bin/activate; make -C u-boot CROSS_COMPILE="aarch64-linux-gnu-" -j $(nproc)
make[1]: Entering directory '/home/nsollars/build/rock64-arch-linux-build/u-boot'
  CHK     include/config/uboot.release
  CHK     include/generated/version_autogenerated.h
  CHK     include/generated/timestamp_autogenerated.h
  UPD     include/generated/timestamp_autogenerated.h
  CHK     include/generated/generic-asm-offsets.h
  CHK     include/generated/asm-offsets.h
  HOSTCC  tools/mkenvimage.o
  HOSTLD  tools/mkenvimage
  HOSTCC  tools/aisimage.o
In file included from ./tools/../lib/libfdt/libfdt.h:10:0,
                 from tools/../include/libfdt.h:1,
                 from tools/fdt_host.h:11,
                 from tools/imagetool.h:24,
                 from tools/aisimage.c:8:
/usr/include/libfdt_env.h:70:30: error: conflicting types for 'fdt64_t'
 typedef uint64_t FDT_BITWISE fdt64_t;
                              ^~~~~~~
In file included from <command-line>:0:0:
././include/libfdt_env.h:19:16: note: previous declaration of 'fdt64_t' was here
 typedef __be64 fdt64_t;
                ^~~~~~~
In file included from ././include/libfdt_env.h:12:0,
                 from <command-line>:0:
/usr/include/libfdt_env.h:90:24: error: expected ')' before 'x'
 static inline uint32_t fdt32_to_cpu(fdt32_t x)
                        ^
././include/compiler.h:71:9: error: expected ')' before '&' token
  ((((x) & 0xff000000) >> 24) | \
         ^
././include/compiler.h:71:23: error: expected ')' before '>>' token
  ((((x) & 0xff000000) >> 24) | \
                       ^
././include/compiler.h:71:30: error: expected ')' before '|' token
  ((((x) & 0xff000000) >> 24) | \
                              ^
/usr/include/libfdt_env.h:94:23: error: expected ')' before 'x'
 static inline fdt32_t cpu_to_fdt32(uint32_t x)
                       ^
././include/compiler.h:71:9: error: expected ')' before '&' token
  ((((x) & 0xff000000) >> 24) | \
         ^
././include/compiler.h:71:23: error: expected ')' before '>>' token
  ((((x) & 0xff000000) >> 24) | \
                       ^
././include/compiler.h:71:30: error: expected ')' before '|' token
  ((((x) & 0xff000000) >> 24) | \
                              ^
/usr/include/libfdt_env.h:99:24: error: expected ')' before 'x'
 static inline uint64_t fdt64_to_cpu(fdt64_t x)
                        ^
././include/compiler.h:76:9: error: expected ')' before '&' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
         ^
././include/compiler.h:76:36: error: expected ')' before '>>' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                    ^
././include/compiler.h:76:43: error: expected ')' before '|' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                           ^
/usr/include/libfdt_env.h:103:23: error: expected ')' before 'x'
 static inline fdt64_t cpu_to_fdt64(uint64_t x)
                       ^
././include/compiler.h:76:9: error: expected ')' before '&' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
         ^
././include/compiler.h:76:36: error: expected ')' before '>>' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                    ^
././include/compiler.h:76:43: error: expected ')' before '|' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                           ^
make[2]: *** [scripts/Makefile.host:116: tools/aisimage.o] Error 1
make[1]: *** [Makefile:1278: tools] Error 2
make[1]: Leaving directory '/home/nsollars/build/rock64-arch-linux-build/u-boot'
make: *** [Makefile.uboot.mk:12: out/u-boot/idbloader.img] Error 2

Nige

Just as an update,

I removed the libfdt header files temporarily from /usr/include ( this might be a bug ) and ran the arch.img build again.  The error I have now is,

aarch64-linux-gnu-ld.bfd: arch/arm/lib/crt0_aarch64_efi.o: relocation R_AARCH64_ABS16 against `EFI_SUBSYSTEM' can not be used when making a shared object
make[3]: *** [scripts/Makefile.lib:382: lib/efi_loader/helloworld.so] Error 1
make[2]: *** [scripts/Makefile.build:425: lib/efi_loader] Error 2
make[1]: *** [Makefile:1278: lib] Error 2
make[1]: Leaving directory '/home/nsollars/build/rock64-arch-linux-build/u-boot'
make: *** [Makefile.uboot.mk:12: out/u-boot/idbloader.img] Error 2

Thanks
Nige

Hmm. I think the next step would be to check if it's an issue upstream.

I tried building mainline u-boot on my Arch Linux machine. With `dtc-1.4.5-1` and `2.29.1-2`, I can run the following (I hope that's the right way to build a similar target...):

```
git clone git://git.denx.de/u-boot.git
cd u-boot
ARCH=arm64 CROSS_COMPILE="aarch64-linux-gnu-" make distclean evb-rk3328_defconfig
ARCH=arm64 CROSS_COMPILE="aarch64-linux-gnu-" make
```

, which seems to work:
```
...
  DTC     arch/arm/dts/rk3399-puma-ddr1600.dtb
  DTC     arch/arm/dts/rk3399-puma-ddr1866.dtb
  DTC     arch/arm/dts/rv1108-evb.dtb
make[2]: 'arch/arm/dts/rk3328-evb.dtb' is up to date.
  SHIPPED dts/dt.dtb
  FDTGREP dts/dt-spl.dtb
  CAT     u-boot-dtb.bin
  COPY    u-boot.bin
  SYM     u-boot.sym
  COPY    u-boot.dtb
  CHK     include/config.h
  CFG     u-boot.cfg
  CFGCHK  u-boot.cfg
```

With the current `dtc-1.4.6-1`, I get this error:

```
  HOSTCC  tools/aisimage.o
In file included from /usr/include/libfdt.h:54:0,
                 from include/image.h:56,
                 from tools/aisimage.c:10:
/usr/include/libfdt_env.h:81:24: error: redefinition of ‘fdt16_to_cpu’
 static inline uint16_t fdt16_to_cpu(fdt16_t x)
                        ^~~~~~~~~~~~
```

So that seems to be an issue with `dtc` and/or `u-boot` (or possibly Arch Linux' packaging).

I'm not sure about the issue that looks like it's `binutils`-related.

alright another issue hit,

make[1]: Leaving directory '/home/nsollars/build/rock64-arch-linux-build/arm-trusted-firmware'
make -C arm-trusted-firmware CROSS_COMPILE=/home/nsollars/crosstools/x-tools8/aarch64-unknown-linux-gnueabi/bin/aarch64-unknown-linux-gnu- PLAT=rk3328 bl31
make[1]: Entering directory '/home/nsollars/build/rock64-arch-linux-build/arm-trusted-firmware'
  CC      drivers/arm/gic/common/gic_common.c
  CC      drivers/arm/gic/v2/gicv2_main.c
  CC      drivers/arm/gic/v2/gicv2_helpers.c
  CC      plat/common/plat_gicv2.c
  CC      plat/rockchip/common/rockchip_gicv2.c
  CC      drivers/arm/cci/cci.c
  CC      drivers/delay_timer/delay_timer.c
  CC      drivers/delay_timer/generic_delay_timer.c
  CC      plat/rockchip/common/drivers/parameter/ddr_parameter.c
  CC      plat/rockchip/common/bl31_plat_setup.c
  CC      plat/rockchip/common/plat_pm.c
  CC      plat/rockchip/common/plat_topology.c
  CC      plat/rockchip/common/aarch64/platform_common.c
  CC      plat/rockchip/rk3328/drivers/pmu/pmu.c
  CC      plat/rockchip/rk3328/drivers/soc/soc.c
  CC      bl31/bl31_main.c
  CC      bl31/interrupt_mgmt.c
  CC      bl31/bl31_context_mgmt.c
bl31/bl31_context_mgmt.c: In function 'cm_get_context_by_mpidr':
bl31/bl31_context_mgmt.c:82:2: warning: 'platform_get_core_pos' is deprecated [-Wdeprecated-declarations]
  return cm_get_context_by_index(platform_get_core_pos(mpidr), security_state);
  ^~~~~~
In file included from bl31/bl31_context_mgmt.c:13:0:
include/plat/common/platform.h:390:14: note: declared here
 unsigned int platform_get_core_pos(unsigned long mpidr) __deprecated;
              ^~~~~~~~~~~~~~~~~~~~~
bl31/bl31_context_mgmt.c: In function 'cm_set_context_by_mpidr':
bl31/bl31_context_mgmt.c:93:2: warning: 'platform_get_core_pos' is deprecated [-Wdeprecated-declarations]
  cm_set_context_by_index(platform_get_core_pos(mpidr),
  ^~~~~~~~~~~~~~~~~~~~~~~
In file included from bl31/bl31_context_mgmt.c:13:0:
include/plat/common/platform.h:390:14: note: declared here
 unsigned int platform_get_core_pos(unsigned long mpidr) __deprecated;
              ^~~~~~~~~~~~~~~~~~~~~
bl31/bl31_context_mgmt.c: In function 'cm_init_context':
bl31/bl31_context_mgmt.c:108:3: warning: 'platform_get_core_pos' is deprecated [-Wdeprecated-declarations]
   cm_init_context_by_index(platform_get_core_pos(mpidr), ep);
   ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from bl31/bl31_context_mgmt.c:13:0:
include/plat/common/platform.h:390:14: note: declared here
 unsigned int platform_get_core_pos(unsigned long mpidr) __deprecated;
              ^~~~~~~~~~~~~~~~~~~~~
  CC      common/runtime_svc.c
  CC      services/arm_arch_svc/arm_arch_svc_setup.c
  CC      services/std_svc/std_svc_setup.c
  CC      lib/el3_runtime/cpu_data_array.c
  CC      lib/el3_runtime/aarch64/context_mgmt.c
  CC      lib/cpus/errata_report.c
  CC      lib/psci/psci_off.c
  CC      lib/psci/psci_on.c
  CC      lib/psci/psci_suspend.c
  CC      lib/psci/psci_common.c
  CC      lib/psci/psci_main.c
  CC      lib/psci/psci_setup.c
  CC      lib/psci/psci_system_off.c
  CC      lib/psci/psci_mem_protect.c
  CC      lib/locks/bakery/bakery_lock_coherent.c
  CC      lib/extensions/spe/spe.c
  CC      common/bl_common.c
  CC      common/tf_log.c
  CC      common/tf_printf.c
  CC      common/tf_snprintf.c
  CC      plat/common/plat_log_common.c
  CC      plat/common/platform_helpers_default.c
  CC      plat/common/aarch64/plat_common.c
plat/common/aarch64/plat_common.c: In function 'plat_get_syscnt_freq2':
plat/common/aarch64/plat_common.c:67:2: warning: 'plat_get_syscnt_freq' is deprecated [-Wdeprecated-declarations]
  unsigned long long freq = plat_get_syscnt_freq();
  ^~~~~~~~
In file included from plat/common/aarch64/plat_common.c:10:0:
include/plat/common/platform.h:43:20: note: declared here
 unsigned long long plat_get_syscnt_freq(void) __deprecated;
                    ^~~~~~~~~~~~~~~~~~~~
  CC      lib/stdlib/abort.c
  CC      lib/stdlib/assert.c
  CC      lib/stdlib/exit.c
  CC      lib/stdlib/mem.c
  CC      lib/stdlib/printf.c
  CC      lib/stdlib/putchar.c
  CC      lib/stdlib/puts.c
  CC      lib/stdlib/sscanf.c
  CC      lib/stdlib/strchr.c
  CC      lib/stdlib/strcmp.c
  CC      lib/stdlib/strlen.c
  CC      lib/stdlib/strncmp.c
  CC      lib/stdlib/strnlen.c
  CC      lib/stdlib/subr_prf.c
  CC      lib/stdlib/timingsafe_bcmp.c
  CC      lib/aarch64/xlat_tables.c
  CC      plat/common/aarch64/plat_psci_common.c
  AS      drivers/console/console.S
  AS      drivers/ti/uart/16550_console.S
  AS      lib/cpus/aarch64/aem_generic.S
  AS      lib/cpus/aarch64/cortex_a53.S
  AS      plat/rockchip/common/aarch64/plat_helpers.S
  AS      plat/rockchip/common/pmusram/pmu_sram_cpus_on.S
  AS      bl31/aarch64/bl31_entrypoint.S
  AS      bl31/aarch64/runtime_exceptions.S
  AS      bl31/aarch64/crash_reporting.S
  AS      plat/common/aarch64/platform_mp_stack.S
  AS      lib/el3_runtime/aarch64/cpu_data.S
  AS      lib/cpus/aarch64/cpu_helpers.S
  AS      lib/locks/exclusive/aarch64/spinlock.S
  AS      lib/psci/aarch64/psci_helpers.S
  AS      lib/el3_runtime/aarch64/context.S
  AS      common/aarch64/debug.S
  AS      lib/aarch64/cache_helpers.S
  AS      lib/aarch64/misc_helpers.S
  AS      plat/common/aarch64/platform_helpers.S
  PP      bl31/bl31.ld.S
  LD      build/rk3328/release/bl31/bl31.elf
./build/rk3328/release/bl31/ddr_parameter.o: In function `ddr_region_usage_parse':
ddr_parameter.c:(.text.ddr_region_usage_parse+0x18): undefined reference to `__stack_chk_guard'
ddr_parameter.c:(.text.ddr_region_usage_parse+0x1c): undefined reference to `__stack_chk_guard'
ddr_parameter.c:(.text.ddr_region_usage_parse+0x6c): undefined reference to `__stack_chk_guard'
ddr_parameter.c:(.text.ddr_region_usage_parse+0x84): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/soc.o: In function `sgrf_init':
soc.c:(.text.sgrf_init+0xc): undefined reference to `__stack_chk_guard'
soc.c:(.text.sgrf_init+0x10): undefined reference to `__stack_chk_guard'
soc.c:(.text.sgrf_init+0xd8): undefined reference to `__stack_chk_guard'
soc.c:(.text.sgrf_init+0xec): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_off.o: In function `psci_do_cpu_off':
psci_off.c:(.text.psci_do_cpu_off+0x14): undefined reference to `__stack_chk_guard'
psci_off.c:(.text.psci_do_cpu_off+0x18): undefined reference to `__stack_chk_guard'
psci_off.c:(.text.psci_do_cpu_off+0x100): undefined reference to `__stack_chk_guard'
psci_off.c:(.text.psci_do_cpu_off+0x118): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_suspend.o: In function `psci_cpu_suspend_start':
psci_suspend.c:(.text.psci_cpu_suspend_start+0xc): undefined reference to `__stack_chk_guard'
psci_suspend.c:(.text.psci_cpu_suspend_start+0x18): undefined reference to `__stack_chk_guard'
psci_suspend.c:(.text.psci_cpu_suspend_start+0x18c): undefined reference to `__stack_chk_guard'
psci_suspend.c:(.text.psci_cpu_suspend_start+0x1a0): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_common.o: In function `psci_release_pwr_domain_locks':
psci_common.c:(.text.psci_release_pwr_domain_locks+0x1c): undefined reference to `__stack_chk_guard'
psci_common.c:(.text.psci_release_pwr_domain_locks+0x20): undefined reference to `__stack_chk_guard'
psci_common.c:(.text.psci_release_pwr_domain_locks+0x68): undefined reference to `__stack_chk_guard'
psci_common.c:(.text.psci_release_pwr_domain_locks+0x7c): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_common.o: In function `psci_warmboot_entrypoint':
psci_common.c:(.text.psci_warmboot_entrypoint+0xc): undefined reference to `__stack_chk_guard'
psci_common.c:(.text.psci_warmboot_entrypoint+0x10): undefined reference to `__stack_chk_guard'
psci_common.c:(.text.psci_warmboot_entrypoint+0xac): undefined reference to `__stack_chk_guard'
psci_common.c:(.text.psci_warmboot_entrypoint+0xc4): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_main.o: In function `psci_cpu_on':
psci_main.c:(.text.psci_cpu_on+0xc): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_cpu_on+0x18): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_cpu_on+0x58): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_cpu_on+0x6c): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_main.o: In function `psci_cpu_suspend':
psci_main.c:(.text.psci_cpu_suspend+0xc): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_cpu_suspend+0x18): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_cpu_suspend+0xa4): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_cpu_suspend+0xbc): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_main.o: In function `psci_system_suspend':
psci_main.c:(.text.psci_system_suspend+0x10): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_system_suspend+0x14): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_system_suspend+0x68): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_system_suspend+0x80): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_main.o: In function `psci_migrate':
psci_main.c:(.text.psci_migrate+0x10): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_migrate+0x14): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_migrate+0x3c): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_migrate+0x50): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_main.o: In function `psci_migrate_info_type':
psci_main.c:(.text.psci_migrate_info_type+0xc): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_migrate_info_type+0x10): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_migrate_info_type+0x38): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_main.o: In function `psci_migrate_info_up_cpu':
psci_main.c:(.text.psci_migrate_info_up_cpu+0xc): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_migrate_info_up_cpu+0x10): undefined reference to `__stack_chk_guard'
psci_main.c:(.text.psci_migrate_info_up_cpu+0x48): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_setup.o: In function `psci_setup':
psci_setup.c:(.text.psci_setup+0x10): undefined reference to `__stack_chk_guard'
psci_setup.c:(.text.psci_setup+0x14): undefined reference to `__stack_chk_guard'
psci_setup.c:(.text.psci_setup+0x26c): undefined reference to `__stack_chk_guard'
psci_setup.c:(.text.psci_setup+0x284): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_mem_protect.o: In function `psci_mem_protect':
psci_mem_protect.c:(.text.psci_mem_protect+0x14): undefined reference to `__stack_chk_guard'
psci_mem_protect.c:(.text.psci_mem_protect+0x1c): undefined reference to `__stack_chk_guard'
psci_mem_protect.c:(.text.psci_mem_protect+0x44): undefined reference to `__stack_chk_guard'
psci_mem_protect.c:(.text.psci_mem_protect+0x58): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/tf_log.o: In function `tf_log':
tf_log.c:(.text.tf_log+0x14): undefined reference to `__stack_chk_guard'
tf_log.c:(.text.tf_log+0x1c): undefined reference to `__stack_chk_guard'
tf_log.c:(.text.tf_log+0x90): undefined reference to `__stack_chk_guard'
tf_log.c:(.text.tf_log+0xa4): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/tf_printf.o: In function `unsigned_num_print':
tf_printf.c:(.text.unsigned_num_print+0x14): undefined reference to `__stack_chk_guard'
tf_printf.c:(.text.unsigned_num_print+0x20): undefined reference to `__stack_chk_guard'
tf_printf.c:(.text.unsigned_num_print+0x8c): undefined reference to `__stack_chk_guard'
tf_printf.c:(.text.unsigned_num_print+0xa0): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/xlat_tables.o: In function `init_xlat_tables':
xlat_tables.c:(.text.init_xlat_tables+0x4): undefined reference to `__stack_chk_guard'
xlat_tables.c:(.text.init_xlat_tables+0x8): undefined reference to `__stack_chk_guard'
xlat_tables.c:(.text.init_xlat_tables+0x68): undefined reference to `__stack_chk_guard'
xlat_tables.c:(.text.init_xlat_tables+0x80): undefined reference to `__stack_chk_fail'
make[1]: *** [Makefile:616: build/rk3328/release/bl31/bl31.elf] Error 1
make[1]: Leaving directory '/home/nsollars/build/rock64-arch-linux-build/arm-trusted-firmware'
make: *** [Makefile.atf.mk:5: arm-trusted-firmware/build/rk3328/release/bl31/bl31.elf] Error 2

Nige

So I decided to give the latest Linaro toolchain a shot and it got quite a bit further but bailed here,

start=$(/home/nsollars/linaro/gcc-linaro-7.2.1-2017.11-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-nm u-boot | grep __rel_dyn_start | cut -f 1 -d ' '); end=$(/home/nsollars/linaro/gcc-linaro-7.2.1-2017.11-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-nm u-boot | grep __rel_dyn_end | cut -f 1 -d ' '); tools/relocate-rela u-boot-nodtb.bin 0x00200000 $start $end
make[3]: 'arch/arm/dts/rk3328-rock64.dtb' is up to date.
./"board/rockchip/evb_rk3328/mk_fit_atf.sh" \
arch/arm/dts/rk3328-rock64.dtb > u-boot.its
ERROR: BL31 file  NOT found, resulting binary is non-functional
make[1]: *** [Makefile:985: u-boot.its] Error 1
make[1]: Leaving directory '/home/nsollars/rock64-arch-linux-build/u-boot'
make: *** [Makefile.uboot.mk:13: out/u-boot/idbloader.img] Error 2
[nsollars@virtualarch rock64-arch-linux-build]$ 

I had a quick look around but did not see anything obvious,  perhaps il delve into the script tomorrow..

Nige

OK, I think I've managed to figure out some things:

### Stack Smashing Protection errors

Reported by @nsollars [here](https://github.com/m01/rock64-arch-linux-build/issues/3#issuecomment-367513256)

```
...
  AS      lib/aarch64/cache_helpers.S
  AS      lib/aarch64/misc_helpers.S
  AS      plat/common/aarch64/platform_helpers.S
  PP      bl31/bl31.ld.S
  LD      build/rk3328/release/bl31/bl31.elf
./build/rk3328/release/bl31/ddr_parameter.o: In function `ddr_region_usage_parse':
ddr_parameter.c:(.text.ddr_region_usage_parse+0x18): undefined reference to `__stack_chk_guard'
ddr_parameter.c:(.text.ddr_region_usage_parse+0x1c): undefined reference to `__stack_chk_guard'
ddr_parameter.c:(.text.ddr_region_usage_parse+0x6c): undefined reference to `__stack_chk_guard'
ddr_parameter.c:(.text.ddr_region_usage_parse+0x84): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/soc.o: In function `sgrf_init':
soc.c:(.text.sgrf_init+0xc): undefined reference to `__stack_chk_guard'
soc.c:(.text.sgrf_init+0x10): undefined reference to `__stack_chk_guard'
soc.c:(.text.sgrf_init+0xd8): undefined reference to `__stack_chk_guard'
soc.c:(.text.sgrf_init+0xec): undefined reference to `__stack_chk_fail'
./build/rk3328/release/bl31/psci_off.o: In function `psci_do_cpu_off':
...
```

I believe these occur because the [Arch Linux ARM pre-build distcc toolchains](https://archlinuxarm.org/wiki/Distcc_Cross-Compiling) are compiled with `--enable-default-ssp`:

```
$ /x-tools8/aarch64-unknown-linux-gnueabi/bin/aarch64-unknown-linux-gnueabi-gcc -v
Using built-in specs.
COLLECT_GCC=/x-tools8/aarch64-unknown-linux-gnueabi/bin/aarch64-unknown-linux-gnueabi-gcc
COLLECT_LTO_WRAPPER=/x-tools8/aarch64-unknown-linux-gnueabi/bin/../libexec/gcc/aarch64-unknown-linux-gnueabi/7.2.1/lto-wrapper
Target: aarch64-unknown-linux-gnueabi
Configured with: /home/leming/cross/bin/.build/src/gcc-7.2.1/configure --build=x86_64-build_pc-linux-gnu --host=x86_64-build_pc-linux-gnu --target=aarch64-unknown-linux-gnueabi --prefix=/home/leming/x-tools8-new/aarch64-unknown-linux-gnueabi --with-sysroot=/home/leming/x-tools8-new/aarch64-unknown-linux-gnueabi/aarch64-unknown-linux-gnueabi/sysroot --enable-languages=c,c++,fortran --with-arch=armv8-a --with-pkgversion='crosstool-NG crosstool-ng-1.23.0-36-g1dbb06f' --enable-__cxa_atexit --disable-libmudflap --enable-libgomp --disable-libssp --enable-libquadmath --enable-libquadmath-support --enable-libsanitizer --disable-libmpx --with-gmp=/home/leming/cross/bin/.build/aarch64-unknown-linux-gnueabi/buildtools --with-mpfr=/home/leming/cross/bin/.build/aarch64-unknown-linux-gnueabi/buildtools --with-mpc=/home/leming/cross/bin/.build/aarch64-unknown-linux-gnueabi/buildtools --with-isl=/home/leming/cross/bin/.build/aarch64-unknown-linux-gnueabi/buildtools --enable-lto --with-host-libstdcxx='-static-libgcc -Wl,-Bstatic,-lstdc++,-Bdynamic -lm' --enable-threads=posix --enable-target-optspace --with-linker-hash-style=gnu --disable-plugin --disable-nls --disable-multilib --with-local-prefix=/home/leming/x-tools8-new/aarch64-unknown-linux-gnueabi/aarch64-unknown-linux-gnueabi/sysroot --enable-long-long --with-arch=armv8-a --enable-default-pie --enable-default-ssp --enable-fix-cortex-a53-835769 --enable-fix-cortex-a53-843419
Thread model: posix
gcc version 7.2.1 20180116 (crosstool-NG crosstool-ng-1.23.0-36-g1dbb06f)
```

Whereas the upstream Arch Linux `aarch64-linux-gnu-gcc` is not:
```
$ aarch64-linux-gnu-gcc -v
Using built-in specs.
COLLECT_GCC=aarch64-linux-gnu-gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/aarch64-linux-gnu/7.3.0/lto-wrapper
Target: aarch64-linux-gnu
Configured with: /build/aarch64-linux-gnu-gcc/src/gcc-7-20180125/configure --prefix=/usr --program-prefix=aarch64-linux-gnu- --with-local-prefix=/usr/aarch64-linux-gnu --with-sysroot=/usr/aarch64-linux-gnu --with-build-sysroot=/usr/aarch64-linux-gnu --libdir=/usr/lib --libexecdir=/usr/lib --target=aarch64-linux-gnu --host=x86_64-pc-linux-gnu --build=x86_64-pc-linux-gnu --disable-nls --enable-languages=c,c++ --enable-shared --enable-threads=posix --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --disable-multilib --disable-werror --enable-checking=release
Thread model: posix
gcc version 7.3.0 (GCC)
```

I believe this causes the Arch Linux ARM pre-built distcc toolchain's gcc to work as if `-fstack-protector-strong` was passed. This means that the https://github.com/ARM-software/arm-trusted-firmware code's `ENABLE_STACK_PROTECTOR` option's default (`"0"`) doesn't actually disable the stack smashing protection. Note that according to the [ARM Trusted Firmware user guide](https://github.com/ARM-software/arm-trusted-firmware/blob/master/docs/user-guide.rst),

> For all values other than 0, the plat_get_stack_protector_canary() platform hook needs to be implemented. 

... Which of course is not the case for the RK3328. Implementing [plat_get_stack_protector_canary](https://github.com/ARM-software/arm-trusted-firmware/blob/master/docs/porting-guide.rst#function-plat-get-stack-protector-canary) might not be too tricky for someone very familiar with the RK3328, but that's probably not the first step to be taken to get Arch Linux to build again for the Rock64.

To work around this issue, we can pass `-fno-stack-protector` to the `CFLAGS` used to build the ARM trusted firmware. I believe this should cause bl31 to be built in a way similar to how it was done before, using the upstream Arch Linux' aarch64 toolchain and the default `ENABLE_STACK_PROTECTOR` setting.

This issue remains when compiling with the current upstream Arch Linux aarch64 cross-compilation toolchain:

```
aarch64-linux-gnu-ld.bfd: arch/arm/lib/crt0_aarch64_efi.o: relocation R_AARCH64_ABS16 against `EFI_SUBSYSTEM' can not be used when making a shared object
make[4]: *** [scripts/Makefile.lib:382: lib/efi_loader/helloworld.so] Error 1
make[3]: *** [/scratch/rock64/archbuild-20180311/rock64-arch-linux-build/u-boot/scripts/Makefile.build:425: lib/efi_loader] Error 2
```

Specifically, `aarch64-linux-gnu-binutils-2.30-1-x86_64.pkg.tar.xz`, which contains:
```
$ aarch64-linux-gnu-ld.bfd -v
GNU ld (GNU Binutils) 2.30
```
is causing the problem.

The code compiles fine with the older version of binutils in the Arch Linux ARM toolchain:
```
.../x-tools8/aarch64-unknown-linux-gnueabi/bin/aarch64-unknown-linux-gnueabi-ld.bfd -v
GNU ld (crosstool-NG crosstool-ng-1.23.0-36-g1dbb06f) 2.29.1
```

--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,14 @@
 UBOOT_DEFCONFIG ?= rock64-rk3328_defconfig
 UBOOT_DIR ?= u-boot
 PYTHON ?= python2
+ATF_PLAT ?= rk3328
+BOARD_TARGET = rock64
+BOARD_CHIP = rk3328
+# TOOLCHAIN ?= aarch64-unknown-linux-gnueabi-
+TOOLCHAIN ?= aarch64-linux-gnu-
 
-arch.img: sync u-boot-venv ArchLinuxARM-aarch64-latest.tar.gz out/u-boot/idbloader.img
-	guestfish -N arch.img=disk:2048M -a out/u-boot/idbloader.img --rw -f create.gfs
+arch.img: sync u-boot-venv ArchLinuxARM-aarch64-latest.tar.gz out/u-boot-rock64/idbloader.img
+	guestfish -N arch.img=disk:2048M -a out/u-boot-rock64/idbloader.img --rw -f create.gfs
 
 arch.img.xz: arch.img
 	xz $<
@@ -30,7 +35,7 @@ rkbin:
 sync: rkbin arm-trusted-firmware ArchLinuxARM-aarch64-latest.tar.gz u-boot
 
 clean:
-	rm -rfv arch.img out
+	rm -rfv arch.img out tmp
 
 superclean: clean
 	rm -rf arm-trusted-firmware u-boot rkbin \

--- a/Makefile.atf.mk
+++ b/Makefile.atf.mk
@@ -2,11 +2,17 @@ BL31 ?= arm-trusted-firmware/build/$(ATF_PLAT)/release/bl31/bl31.elf
 
 arm-trusted-firmware/build/$(ATF_PLAT)/release/bl31/bl31.elf: arm-trusted-firmware
 	make -C $< realclean
-	make -C $< CROSS_COMPILE=aarch64-linux-gnu- M0_CROSS_COMPILE=arm-linux-gnueabi- PLAT=$(ATF_PLAT) bl31
+	# Note that ENABLE_STACK_PROTECTOR="0" is actually the default,
+	# see https://github.com/ARM-software/arm-trusted-firmware/blob/master/docs/user-guide.rst
+	# However, if $(TOOLCHAIN)gcc has been compiled with --enable-default-ssp,
+	# we need to explicitly disable the stack protector, since currently
+	# plat_get_stack_protector_canary() is not implemented for the Rock64.
+	# See https://github.com/m01/rock64-arch-linux-build/issues/3 for details.
+	make -C $< CROSS_COMPILE=$(TOOLCHAIN) M0_CROSS_COMPILE=$(TOOLCHAIN) PLAT=$(ATF_PLAT) CFLAGS="-fno-stack-protector" bl31 ENABLE_STACK_PROTECTOR="0"
 
 arm-trusted-firmware/build/$(ATF_PLAT)/debug/bl31/bl31.elf: arm-trusted-firmware
 	make -C $< realclean
-	make -C $< CROSS_COMPILE=aarch64-linux-gnu- M0_CROSS_COMPILE=arm-linux-gnueabi- PLAT=$(ATF_PLAT) bl31 DEBUG=1
+	make -C $< CROSS_COMPILE=$(TOOLCHAIN) M0_CROSS_COMPILE=$(TOOLCHAIN) PLAT=$(ATF_PLAT) CFLAGS="-fno-stack-protector" bl31 DEBUG=1 ENABLE_STACK_PROTECTOR="0"
 
 .PHONY: atf-build
 atf-build: $(BL31)

--- a/Makefile.atf.mk
+++ b/Makefile.atf.mk
@@ -1,12 +1,12 @@
-BL31 ?= arm-trusted-firmware/build/rk3328/release/bl31/bl31.elf
+BL31 ?= arm-trusted-firmware/build/$(ATF_PLAT)/release/bl31/bl31.elf
 
-arm-trusted-firmware/build/rk3328/release/bl31/bl31.elf: arm-trusted-firmware
+arm-trusted-firmware/build/$(ATF_PLAT)/release/bl31/bl31.elf: arm-trusted-firmware
 	make -C $< realclean
-	make -C $< CROSS_COMPILE=aarch64-linux-gnu- PLAT=rk3328 bl31
+	make -C $< CROSS_COMPILE=aarch64-linux-gnu- M0_CROSS_COMPILE=arm-linux-gnueabi- PLAT=$(ATF_PLAT) bl31
 
-arm-trusted-firmware/build/rk3328/debug/bl31/bl31.elf: arm-trusted-firmware
+arm-trusted-firmware/build/$(ATF_PLAT)/debug/bl31/bl31.elf: arm-trusted-firmware
 	make -C $< realclean
-	make -C $< CROSS_COMPILE=aarch64-linux-gnu- PLAT=rk3328 bl31 DEBUG=1
+	make -C $< CROSS_COMPILE=aarch64-linux-gnu- M0_CROSS_COMPILE=arm-linux-gnueabi- PLAT=$(ATF_PLAT) bl31 DEBUG=1
 
 .PHONY: atf-build
 atf-build: $(BL31)

--- a/Makefile.uboot.mk
+++ b/Makefile.uboot.mk
@@ -1,64 +1,64 @@
-UBOOT_MAKE ?= source $(UBOOT_DIR)/../u-boot-venv/bin/activate; make -C $(UBOOT_DIR) \
-	CROSS_COMPILE="aarch64-linux-gnu-"
+UBOOT_OUTPUT_DIR ?= $(CURDIR)/tmp/u-boot-$(BOARD_TARGET)
+UBOOT_LOADER ?= out/u-boot-$(BOARD_TARGET)/idbloader.img
+UBOOT_MAKE ?= make -C $(UBOOT_DIR) KBUILD_OUTPUT=$(UBOOT_OUTPUT_DIR) BL31=$(realpath $(BL31)) \
+	CROSS_COMPILE="ccache aarch64-linux-gnu-"
 
-$(UBOOT_DIR)/.config: $(UBOOT_DIR)/configs/$(UBOOT_DEFCONFIG)
-	$(UBOOT_MAKE) $(UBOOT_DEFCONFIG)
+tmp/u-boot-$(BOARD_TARGET)/.config: $(UBOOT_DIR)/configs/$(UBOOT_DEFCONFIG)
+	$(UBOOT_MAKE) $(UBOOT_DEFCONFIG) 
 
-ifneq (,$(FORCE))
-.PHONY: out/u-boot/idbloader.img
-endif
-out/u-boot/idbloader.img: $(UBOOT_DIR)/.config $(BL31) u-boot-venv u-boot
-	cp -u $(BL31) u-boot/bl31.elf
+$(UBOOT_LOADER): tmp/u-boot-$(BOARD_TARGET)/.config $(BL31)
 	$(UBOOT_MAKE) -j $$(nproc)
 	$(UBOOT_MAKE) -j $$(nproc) u-boot.itb
-	mkdir -p out/u-boot
-ifneq (,$(USE_UBOOT_TPL))
-	$(UBOOT_DIR)/tools/mkimage -n rk3328 -T rksd -d $(UBOOT_DIR)/tpl/u-boot-tpl.bin $@.tmp
+	mkdir -p out/u-boot-$(BOARD_TARGET)
+ifneq (,$(USE_UBOOT_SPL))
+	$(UBOOT_OUTPUT_DIR)/tools/mkimage -n $(BOARD_CHIP) -T rksd -d $(UBOOT_OUTPUT_DIR)/spl/u-boot-spl.bin $@.tmp
+else ifneq (,$(USE_UBOOT_TPL))
+	$(UBOOT_OUTPUT_DIR)/tools/mkimage -n $(BOARD_CHIP) -T rksd -d $(UBOOT_OUTPUT_DIR)/tpl/u-boot-tpl.bin $@.tmp
+	cat $(UBOOT_OUTPUT_DIR)/spl/u-boot-spl.bin >> $@.tmp
+else ifeq (rock64,$(BOARD_TARGET))
+	$(UBOOT_OUTPUT_DIR)/tools/mkimage -n $(BOARD_CHIP) -T rksd -d rkbin/rk33/rk3328_ddr_786MHz_v1.06.bin $@.tmp
+	cat $(UBOOT_OUTPUT_DIR)/spl/u-boot-spl.bin >> $@.tmp
+else ifeq (rockpro64,$(BOARD_TARGET))
+	$(UBOOT_OUTPUT_DIR)/tools/mkimage -n $(BOARD_CHIP) -T rksd -d rkbin/rk33/rk3399_ddr_800MHz_v1.08.bin $@.tmp
+	cat $(UBOOT_OUTPUT_DIR)/spl/u-boot-spl.bin >> $@.tmp
 else
-	$(UBOOT_DIR)/tools/mkimage -n rk3328 -T rksd -d rkbin/rk33/rk3328_ddr_786MHz_v1.06.bin $@.tmp
+	@echo "Invalid $(BOARD_TARGET)"
+	@exit 1
 endif
-	cat $(UBOOT_DIR)/spl/u-boot-spl.bin >> $@.tmp
-	dd if=$(UBOOT_DIR)/u-boot.itb of=$@.tmp seek=$$((0x200-64)) conv=notrunc
+	dd if=$(UBOOT_OUTPUT_DIR)/u-boot.itb of=$@.tmp seek=$$((0x200-64)) conv=notrunc
 	mv $@.tmp $@
 
 .PHONY: u-boot-menuconfig		# edit u-boot config and save as defconfig
-u-boot-menuconfig: u-boot-venv u-boot
+u-boot-menuconfig:
 	$(UBOOT_MAKE) ARCH=arm64 $(UBOOT_DEFCONFIG)
 	$(UBOOT_MAKE) ARCH=arm64 menuconfig
 	$(UBOOT_MAKE) ARCH=arm64 savedefconfig
-	cp $(UBOOT_DIR)/defconfig $(UBOOT_DIR)/configs/$(UBOOT_DEFCONFIG)
+	mv $(UBOOT_OUTPUT_DIR)/defconfig $(UBOOT_DIR)/configs/$(UBOOT_DEFCONFIG)
+
+.PHONY: u-boot-build		# compile u-boot
+u-boot-build:
+	rm -f $(UBOOT_LOADER)
+	make $(UBOOT_LOADER)
 
 .PHONY: u-boot-clear
 u-boot-clear:
-	rm -rf out/u-boot/
+	rm -rf $(UBOOT_LOADER)/..
 
-.PHONY: u-boot-boot		# boot u-boot over USB
-u-boot-boot: out/u-boot/idbloader.img
-	rkdeveloptool db rkbin/rk33/rk3328_loader_v1.08.244_for_spi_nor_build_Aug_7_2017.bin
-	sleep 1s
-	rkdeveloptool rid
-	rkdeveloptool wl 512 $(UBOOT_DIR)/u-boot.itb
-	rkdeveloptool rd
-	sleep 1s
+out/u-boot/%/boot.scr: blobs/%.cmd
+	mkdir -p $$(dirname $@)
+	mkimage -C none -A arm -T script -d $< $@
 
-	cat rkbin/rk33/rk3328_ddr_786MHz_v1.06.bin | openssl rc4 -K 7c4e0304550509072d2c7b38170d1711 | rkflashtool l
-	cat u-boot/spl/u-boot-spl.bin | openssl rc4 -K 7c4e0304550509072d2c7b38170d1711 | rkflashtool L
+out/u-boot/%/boot.img: out/u-boot/%/boot.scr
+	dd if=/dev/zero of=$@ bs=1M count=2
+	mkfs.vfat -n "u-boot-script" $@
+	mcopy -sm -i $@ $< ::
 
-.PHONY: u-boot-flash-spi		# flash u-boot to SPI
-u-boot-flash-spi: out/u-boot/idbloader.img
-	rkdeveloptool db rkbin/rk33/rk3328_loader_v1.08.244_for_spi_nor_build_Aug_7_2017.bin
-	sleep 1s
-	rkdeveloptool rid
-	rkdeveloptool wl 64 $<
-	rkdeveloptool rd
+u-boot-%-$(BOARD_TARGET).img: out/u-boot/%/boot.img $(UBOOT_LOADER)
+	build/mk-image.sh -c $(BOARD_CHIP) -d out/u-boot-$(BOARD_TARGET) -t system -s 128 -b $< -o "$@.tmp"
+	mv "$@.tmp" $@
 
-.PHONY: u-boot-clear-spi		# clear u-boot to SPI
-u-boot-clear-spi: out/u-boot/idbloader.img
-	rkdeveloptool db rkbin/rk33/rk3328_loader_v1.08.244_for_spi_nor_build_Aug_7_2017.bin
-	sleep 1s
-	rkdeveloptool rid
-	rkdeveloptool wl 64 $<
-	rkdeveloptool rd
+.PHONY: u-boot-flash-spi-$(BOARD_TARGET)
+u-boot-flash-spi-$(BOARD_TARGET): u-boot-flash-spi-$(BOARD_TARGET).img.xz
 
-.PHONY: u-boot-build		# compile u-boot
-u-boot-build: out/u-boot/idbloader.img
+.PHONY: u-boot-erase-spi-$(BOARD_TARGET)
+u-boot-erase-spi-$(BOARD_TARGET): u-boot-erase-spi-$(BOARD_TARGET).img.xz

--- a/Makefile.uboot.mk
+++ b/Makefile.uboot.mk
@@ -1,12 +1,12 @@
 UBOOT_OUTPUT_DIR ?= $(CURDIR)/tmp/u-boot-$(BOARD_TARGET)
 UBOOT_LOADER ?= out/u-boot-$(BOARD_TARGET)/idbloader.img
 UBOOT_MAKE ?= make -C $(UBOOT_DIR) KBUILD_OUTPUT=$(UBOOT_OUTPUT_DIR) BL31=$(realpath $(BL31)) \
-	CROSS_COMPILE="ccache aarch64-linux-gnu-"
+	CROSS_COMPILE="$(TOOLCHAIN)"
 
-tmp/u-boot-$(BOARD_TARGET)/.config: $(UBOOT_DIR)/configs/$(UBOOT_DEFCONFIG)
-	$(UBOOT_MAKE) $(UBOOT_DEFCONFIG) 
+tmp/u-boot-$(BOARD_TARGET)/.config: $(UBOOT_DIR)/configs/$(UBOOT_DEFCONFIG) u-boot-venv u-boot
+	$(UBOOT_MAKE) $(UBOOT_DEFCONFIG)
 
-$(UBOOT_LOADER): tmp/u-boot-$(BOARD_TARGET)/.config $(BL31)
+$(UBOOT_LOADER): tmp/u-boot-$(BOARD_TARGET)/.config $(BL31) u-boot-venv u-boot
 	$(UBOOT_MAKE) -j $$(nproc)
 	$(UBOOT_MAKE) -j $$(nproc) u-boot.itb
 	mkdir -p out/u-boot-$(BOARD_TARGET)
@@ -29,14 +29,14 @@ endif
 	mv $@.tmp $@
 
 .PHONY: u-boot-menuconfig		# edit u-boot config and save as defconfig
-u-boot-menuconfig:
+u-boot-menuconfig: u-boot-venv u-boot
 	$(UBOOT_MAKE) ARCH=arm64 $(UBOOT_DEFCONFIG)
 	$(UBOOT_MAKE) ARCH=arm64 menuconfig
 	$(UBOOT_MAKE) ARCH=arm64 savedefconfig
 	mv $(UBOOT_OUTPUT_DIR)/defconfig $(UBOOT_DIR)/configs/$(UBOOT_DEFCONFIG)
 
 .PHONY: u-boot-build		# compile u-boot
-u-boot-build:
+u-boot-build: u-boot-venv u-boot
 	rm -f $(UBOOT_LOADER)
 	make $(UBOOT_LOADER)
 
@@ -44,16 +44,16 @@ u-boot-build:
 u-boot-clear:
 	rm -rf $(UBOOT_LOADER)/..
 
-out/u-boot/%/boot.scr: blobs/%.cmd
+out/u-boot/%/boot.scr: blobs/%.cmd u-boot-venv u-boot
 	mkdir -p $$(dirname $@)
 	mkimage -C none -A arm -T script -d $< $@
 
-out/u-boot/%/boot.img: out/u-boot/%/boot.scr
+out/u-boot/%/boot.img: out/u-boot/%/boot.scr u-boot-venv u-boot
 	dd if=/dev/zero of=$@ bs=1M count=2
 	mkfs.vfat -n "u-boot-script" $@
 	mcopy -sm -i $@ $< ::
 
-u-boot-%-$(BOARD_TARGET).img: out/u-boot/%/boot.img $(UBOOT_LOADER)
+u-boot-%-$(BOARD_TARGET).img: out/u-boot/%/boot.img $(UBOOT_LOADER) u-boot-venv u-boot
 	build/mk-image.sh -c $(BOARD_CHIP) -d out/u-boot-$(BOARD_TARGET) -t system -s 128 -b $< -o "$@.tmp"
 	mv "$@.tmp" $@
 

--- a/README.md
+++ b/README.md
@@ -25,6 +25,8 @@ And one of the following toolchains for cross-compilation (see issue #3):
 OR:
 * ARMv8 pre-built toolchain from https://archlinuxarm.org/wiki/Distcc_Cross-Compiling
 
+Ensure the `TOOLCHAIN` variable in the `Makefile` is set appropriately.
+
 Then:
 ```
 make [arch.img]

--- a/README.md
+++ b/README.md
@@ -12,13 +12,18 @@ Build procedure
 This is written from the perspective of someone building the image on Arch Linux x86_64.
 
 Install the following prerequisites (mostly required for building u-boot):
-* `aarch64-linux-gnu-gcc`
-* `guestfish`
-* `dtc`
+* `guestfish` (part of `libguestfs` package)
+* `dtc` version `1.4.5-1`
 * `git`
 * `uboot-tools`
 * `swig`
 * `make`
+
+And one of the following toolchains for cross-compilation (see issue #3):
+* `aarch64-linux-gnu-gcc`, and
+* `aarch64-linux-gnu-binutils` version `2.29.1-1` (tested) or `>=2.31` (presumed). `2.30` does not work, see the issue mentioned above.
+OR:
+* ARMv8 pre-built toolchain from https://archlinuxarm.org/wiki/Distcc_Cross-Compiling
 
 Then:
 ```
@@ -55,3 +60,4 @@ Credits
 * Rockchip for caring about mainline support for their chips
 * Arch Linux ARM for the distribution
 * [Arch Linux Wiki page on banana pi](https://wiki.archlinux.org/index.php/Banana_Pi) for a pointer on fixing python library loading issues when building u-boot
+* Everyone who's reported issues and helped me fix them!

